File: minReorderRoutes.py
from collections import deque
class Solution:
# Approach 1- DFS
    def minReorder(self, n, connections): # TC O(n) // SC O(n)
        
        def dfs(node, parent):
            nonlocal answer
            
            if node not in adj: return 
            
            for neighbor in adj[node]:
                child=neighbor[0]
                sign=neighbor[1]
                if child!=parent:
                    answer+=sign
                    dfs(child, node)
        
        adj={}
        answer=0
        
#         for conn in connections:
#             if conn[0] not in adj:
#                 conn[0]=[]
#             if conn[1] not in adj:
#                 conn[1]=[]
            
#             adj[conn[0]].append([conn[1], 1])
#             adj[conn[1]].append([conn[0],0])
#         adj = {}

        for conn in connections:
            adj.setdefault(conn[0], []).append([conn[1], 1])
            adj.setdefault(conn[1], []).append([conn[0], 0])
        
        dfs(0,-1)
        return answer
        
# Approach 2- BFS
    def minReorder1(self, n, connections): # TC O(n) // SC O(n)

        def bfs(node):
            nonlocal answer
            visit=[False]*(n+1)
            visit[node]=True
            q=deque()
            q.append(node)
            while q:
                node=q.popleft()
                if node not in adj: continue 
                
                for child,sign in adj[node]:
                    if not visit[child]:
                        answer+=sign
                        visit[child]=True
                        q.append(child)
            
            
            return
                  
        adj={}
        answer=0
        
#         for conn in connections:
#             if conn[0] not in adj:
#                 conn[0]=[]
#             if conn[1] not in adj:
#                 conn[1]=[]
            
#             adj[conn[0]].append([conn[1], 1])
#             adj[conn[1]].append([conn[0],0])
#         adj = {}

        for conn in connections:
            adj.setdefault(conn[0], []).append([conn[1], 1])
            adj.setdefault(conn[1], []).append([conn[0], 0])
        
        bfs(0)
        return answer 


s=Solution()
print(s.minReorder(6,[[0,1],[1,3],[2,3],[4,0],[4,5]]))
print(s.minReorder1(6,[[0,1],[1,3],[2,3],[4,0],[4,5]]))

File: minReorderRoutes.java
import java.util.*;
// Approach 1- DFS
class Solution { // TC O(n) SC O(n)
    int count = 0;

    public void dfs(int node, int parent, Map<Integer, List<List<Integer>>> adj) {
        if (!adj.containsKey(node)) {
            return;
        }
        for (List<Integer> nei : adj.get(node)) {
            int child = nei.get(0);
            int sign = nei.get(1);
            if (child != parent) {
                count += sign;
                dfs(child, node, adj);
            }
        }
    }

    public int minReorder(int n, int[][] connections) {
        Map<Integer, List<List<Integer>>> adj = new HashMap<>();
        for (int[] connection : connections) {
            adj.computeIfAbsent(connection[0], k -> new ArrayList<List<Integer>>()).add(
                    Arrays.asList(connection[1], 1));
            adj.computeIfAbsent(connection[1], k -> new ArrayList<List<Integer>>()).add(
                    Arrays.asList(connection[0], 0));
        }
        dfs(0, -1, adj);
        return count;
    }
}

// Approach 2- BFS
class Solution1{ // TC O(n) SC O(n)
    int count = 0;
    public void bfs(int node, int n, Map<Integer, List<List<Integer>>> adj) {
        Queue<Integer> q = new LinkedList<>();
        boolean[] visit = new boolean[n];
        q.offer(node);
        visit[node] = true;

        while (!q.isEmpty()) {
            node = q.poll();
            if (!adj.containsKey(node)) {
                continue;
            }
            for (List<Integer> nei : adj.get(node)) {
                int neighbor = nei.get(0);
                int sign = nei.get(1);
                if (!visit[neighbor]) {
                    count += sign;
                    visit[neighbor] = true;
                    q.offer(neighbor);
                }
            }
        }
    }

    public int minReorder(int n, int[][] connections) {
        Map<Integer, List<List<Integer>>> adj = new HashMap<>();
        for (int[] connection : connections) {
            adj.computeIfAbsent(connection[0], k -> new ArrayList<List<Integer>>()).add(
                    Arrays.asList(connection[1], 1));
            adj.computeIfAbsent(connection[1], k -> new ArrayList<List<Integer>>()).add(
                    Arrays.asList(connection[0], 0));
        }
        bfs(0, n, adj);
        return count;
    }
}

class minReorderRoutes{
    public static void main(String[] args) {
        int[][] connections={{0,1},{1,3},{2,3},{4,0},{4,5}};
        int n=6;
        Solution s= new Solution();
        System.out.println(s.minReorder(n, connections));
        Solution1 s1= new Solution1();
        System.out.println(s1.minReorder(n, connections));
    }
}